Improvement the validation of the eav entity attribute

eduard13 opened the pull request:
### Description
This additional validation, is checking for allowed characters within an eav attribute code, which allows only letters, number, underscores and the first character should be a letter.


### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Improvement the validation of the eav entity attribute

magento-engcom-team commented:
Hi @eduard13. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on Pull Request changes
- `@magento-engcom-team give me new test instance` - deploy **NEW** test instance based on Pull Request changes
- `@magento-engcom-team give me {$VERSION} instance` - deploy Vanilla Magento instance for Issue or Pull Request

For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation

adeptofvoltron commented on app/code/Magento/Eav/Model/Entity/Attribute.php:
```diff
@@ -253,6 +253,16 @@ public function beforeSave()
+         * Check attribute_code for allowed characters
+         */
+        if (isset($this->_data['attribute_code'])
+            && !preg_match('/^[a-z]+[a-z0-9_]+$/', $this->_data['attribute_code'])) {
```
that regex, also force us to have attribute_code with at least 2 chars. Maybe it is ok, I just want to make you aware of that.

orlangur commented on app/code/Magento/Eav/Model/Entity/Attribute.php:
```diff
@@ -253,6 +253,16 @@ public function beforeSave()
+         * Check attribute_code for allowed characters
+         */
+        if (isset($this->_data['attribute_code'])
+            && !preg_match('/^[a-z]+[a-z0-9_]+$/', $this->_data['attribute_code'])) {
```
Model is not right place for validation, there is some validation already present in controller, isn't it?

eduard13 replied:
@orlangur I suggest to have this validation also when we need to create some eav attributes programmatically. As for the controllers, there is a frontend validation in the backend while an attribute is created.

orlangur replied:
@eduard13 frontend validation is not enough. I'm pretty sure this is validated on PHP side also. Validation should not be added to model itself, when you create attribute programmatically - you should validate attribute code prior to passing it into model.

eduard13 replied:
@orlangur I agree with you that we should check the attribute code before saving it, but sometimes you don't know that some chars aren't allowed, like . (dots), and you successfully save the attribute, but later you aren't able to properly work with the attribute (because the getters and setters aren't well generated).

By the way, if we'll take a look closer in this method, there are other several validation rules, so it is needed to refactor this method and move all the validations to a separate file?

orlangur replied:
@eduard13 didn't check the code out of your changeset :) I need some time to understand the whole picture, please be patient.

eduard13 replied:
Hi @orlangur any suggestions here?

Commit:
Fixed line exceeds maximum limit

orlangur committed:
Improve error message wording

orlangur reviewed (changes requested):
@eduard13 just discovered that there is already validation for attribute code length here, so, checking for allowed characters should be fine.

Please introduce an `AttributeCodeValidator` class and reuse it here and in `\Magento\Catalog\Model\Product\Attribute\Repository::generateCode` and in `\Magento\Catalog\Model\Product\Attribute\Repository::validateCode`.

Commit:
Merge branch '2.2-develop' into 2.2-develop-eav-attribute-validation

Commit:
Merge remote-tracking branch 'origin/2.2-develop' into 2.2-develop-eav-attribute-validation

Commit:
Revert "Improve error message wording"

This reverts commit 53c120901e3354f8c6ce59773076e4d8ddbffb7b.

Commit:
Revert "Fixed line exceeds maximum limit"

This reverts commit 3f0e34144ee7de64560b47808fb1a4792c729586.

Commit:
Revert "Improvement the validation of the eav entity attribute"

This reverts commit 9df8593395b1db185ce78323eda8b0dd18292bb4.

Commit:
Improving eav attribute validation when creating using setup services.

eduard13 commented:
@orlangur, after digging deeper, I've noticed that when the attributes are created using Setup services, the Attribute model isn't used for saving it, instead of that it is inserted directly in DB. So I've moved the validation in the validating method and reverted the previously made changes.

Moreover, I don't see the reason to use a different validation message, than it is used elsewhere.

Commit:
Fixed failing test regarding invalid attribute code

Commit:
Revert "Fixed failing test regarding invalid attribute code"

This reverts commit 4e2ba807d2f6b079f104290a5f95d2c8b03bdfe5.

Commit:
Adding a separate test for testing the new added validation rule

Commit:
Fixing wrong test attributes

sidolov commented:
Hi @eduard13 , will you continue progress with PR?

orlangur commented:
@sidolov we need a reply from @VladimirZaets first in order to understand a proper validation rule.

magento-engcom-team commented:
Hi @sivaschenko, thank you for the review.
[ENGCOM-3442](https://jira.corp.magento.com/browse/ENGCOM-3442) has been created to process this Pull Request

sivaschenko commented:
Hi @eduard13, it's preferable to merge changes to 2.3 before merging to 2.2 on the current stage. Please [port]((please see [contributor guide](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#rules) for details)) the pull request to 2.3-develop branch

eduard13 commented:
Hi @sivaschenko, thank you. I agree with that, as this PR was made when the mainline was 2.2. Will do it asap.

orlangur reviewed (changes requested):
@eduard13 review NOT passed, please pay attention to https://github.com/magento/magento2/pull/16204#issuecomment-403433664. Sergey is right about delivering everything via `2.3-develop` first nowadays, please create such a PR containing necessary changes.

@sivaschenko please DO NOT touch labels when you don't know what are they intended for.

eduard13 commented:
@sivaschenko we can close this one, I will open a new one for 2.3 with the mentioned reviews fixed.
Thanks everyone.
